CCv0: Refactor CCv0.sh to use integration test lib.sh

Megan-Wright opened the pull request:
Updates ccv0.sh to remove duplicated code
Fixes: #3898
Signed-off-by: Megan Wright <megan.wright@ibm.com>
Signed-off-by: Georgina Kinge  <georgina.kinge@ibm.com>

wainersm commented on docs/how-to/ccv0.sh:
```diff
@@ -51,6 +51,10 @@ export CONTAINER_ID="${CONTAINER_ID:-0123456789}"
 source /etc/os-release || source /usr/lib/os-release
 grep -Eq "\<fedora\>" /etc/os-release 2> /dev/null && export USE_PODMAN=true
 
+if [ -d ${tests_repo_dir} ]; then
```
I'm not sure this checking is needed. Before some variables that holds file/directory path aren't checked. Is there anything special with `BATS_TEST_DIRNAME` that requires this checking?

wainersm commented on docs/how-to/ccv0.sh:
```diff
@@ -51,6 +51,10 @@ export CONTAINER_ID="${CONTAINER_ID:-0123456789}"
 grep -Eq "\<fedora\>" /etc/os-release 2> /dev/null && export USE_PODMAN=true
 
+if [ -d ${tests_repo_dir} ]; then
+    export BATS_TEST_DIRNAME="${tests_repo_dir}/integration/containerd/confidential/"
```
The name `BATS_TESTS_DIRNAME` is a special variable of bats (see [here](https://bats-core.readthedocs.io/en/stable/writing-tests.html#special-variables)). It might silently introduce a bug If in the future `ccv0.sh` calls bats due to clash on the variable names.

wainersm commented on docs/how-to/ccv0.sh:
Really want this line commented?

wainersm replied:
ah, yes, later you call `switch_image_service_offload` so maybe you just need to delete that line.

wainersm commented on docs/how-to/ccv0.sh:
```diff
@@ -435,45 +449,27 @@ crictl_create_cc_pod() {
-metadata:
-  name: ${crictl_sandbox_name}
-EOF
+    source ${tests_repo_dir}/integration/containerd/confidential/lib.sh
```
Likewise.

wainersm commented on docs/how-to/ccv0.sh:
```diff
@@ -435,45 +449,27 @@ crictl_create_cc_pod() {
-        sudo crictl stop ${container_id} && sudo crictl rm ${container_id}
-    fi
-    sudo crictl stopp ${pod_id} && sudo crictl rmp ${pod_id}
+    source ${tests_repo_dir}/integration/containerd/confidential/lib.sh
```
and here.

wainersm commented:
> @wainersm Thanks so much! I've updated almost everything you mentioned and retested to ensure the script still works as expected... the reason we set `$BATS_TEST_DIRNAME` is because it is required to already be set by `lib.sh` in the tests repo when it is called (the code where this variable gets set in testing isn't called/used by us here).

Oh, true, I forgot `lib.sh` was dependent of a `$BATS_TEST_DIRNAME`. Because I source the file from inside a bats test, the variable will be already set by the framework.

wainersm commented:
@Megan-Wright hello, you will need to change your commit to fix the failure on the static check.

jodh-intel commented on docs/how-to/ccv0.sh:
Can you replace the 4 function-local `source` calls with a single top-level `source` maybe around line 51 where we source `/etc/os-release`?

Megan-Wright commented:
Hi @jodh-intel thanks so much for the review! I've addressed all your comments as far as I can - we still need to do the `source ${BATS_TEST_DIRNAME}/lib.sh` in two places as we can't source the code until it's been checked out in `check_out_repos` but I've removed the remaining source commands.

jodh-intel commented on docs/how-to/ccv0.sh:
```diff
@@ -175,6 +183,9 @@ check_out_repos() {
     fi
     git reset --hard origin/${tests_branch}
+
+    source "${BATS_TEST_DIRNAME}/lib.sh"
```
I think this line can be removed?

Megan-Wright replied:
The reason we need to source `lib.sh` in this function as well as globally is because the repo is checked out in this function and the code in `lib.sh` is then referenced before exiting the function :)

stevenhorsman replied:
Hey James, thanks so much for the reviews. As Megan says I think we still need both of the `source`s as they run under different circumstances, depending on whether the code has already been checked out or not:
- The first one: `[ -d "${BATS_TEST_DIRNAME}" ] && source "${BATS_TEST_DIRNAME}/lib.sh"` will only run if the code has already been checked out (otherwise it isn't there to source!)
- This one runs in `check_out_repos()`, which will run for a fresh environment when the script hasn't previously run, but not be run most of the time when the tests repo has already been checked out
Together they should provide coverage of the two different states. I hope this makes sense. If you have a better suggestion to solve this then that would be great though!

jodh-intel commented on docs/how-to/ccv0.sh:
Are these `add_kernel_params()` calls missing an arg?

Megan-Wright replied:
ah yep good spot - I completely missed that... I've updated now

jodh-intel reviewed (commented):
Thanks @Megan-Wright. A few more comments...

Megan-Wright commented on docs/how-to/ccv0.sh:
```diff
@@ -51,6 +51,14 @@ export CONTAINER_ID="${CONTAINER_ID:-0123456789}"
+# If we've already checked out the test repo then source the confidential scripts
+if [ -d "${BATS_TEST_DIRNAME}" ]; then
+    source "${BATS_TEST_DIRNAME}/lib.sh"
+fi
```
done :)

stevenhorsman commented:
/test

stevenhorsman commented:
It's worth noting that in order to work with `ccv0.sh` these changes require https://github.com/kata-containers/tests/pull/4591 to be merged (basically allowing a non-root user to run the integration script methods that this PR is now using), so I've added the `do-not-merge` label.

stevenhorsman commented:
https://github.com/kata-containers/tests/pull/4591 has merged now, so I've removed the `do-not-merge` label.

stevenhorsman commented:
/test

c3d commented on docs/how-to/ccv0.sh:
```diff
@@ -430,50 +419,29 @@ kubernetes_delete_ssh_demo_pod() {
 }
 
-crictl_create_cc_container() {
+unshadow_crictl_create_cc_container() {
```
The `unshadow` prefix is confusing. I think that `unshadow` here refers to the same name in `lib.sh`, but you are not really unshadowing it, more like wrapping it to make it easier to use. It's also a bit inconsistent with the `crictl_delete_cc` case. What about a `wrap` or `call` prefix?

Megan-Wright commented:
@c3d Thanks for the review :) You're right it is confusing, I just couldn't think of a better way to put it! I've updated the prefixes to `call_` now

stevenhorsman commented:
/test

stevenhorsman commented:
@jodh-intel, @wainersm, @c3d - thanks for all the review comments. I think Megan has address them all now, so is there anything else you'd like to see considered/changed before we get this merged?

jodh-intel reviewed (approved):
Thanks @Megan-Wright. I personally still find the "double `source`" is a bit confusing, but other than that...

lgtm
